src/pip/_internal/resolution/resolvelib/resolver.py

```diff
@@ -159,6 +139,34 @@ def get_installation_order(self, req_set):
         return [ireq for _, ireq in sorted_items]
 
 
+def get_topological_weights(graph):
+    # type: (Graph) -> Dict[Optional[str], int]
+    """Assign weights to each node based on how "deep" they are.
+    """
+    visited = set()  # type: Set[Optional[str]]
+    weights = {}
+
+    def visit(node, depth=0):
+        # type: (Optional[str], int) -> None
+        if node in visited:
+            return
+
+        visited.add(node)
+        for child in graph.iter_children(node):
+            visit(child, depth + 1)
+
+        weights[node] = depth
```

Comment:
This feels wrong to me. Say I have (`->` means *depends on*)

```
B -> A
E -> D -> C -> A
```

Then we’d do

```
visit(B, 0)
visit(A, 1)
visit(E, 0)
visit(D, 1)
visit(C, 2)
visit(A, 3)  # But A is visited so we skip it.
```

and then C (weight 2) is installed before A (with weight 1), which is wrong.

Reply:
Indeed. The resulting weight mapping for this graph is:

```
{'A': 2, 'B': 1, 'C': 3, 'D': 2, 'E': 1, None: 0}
```

`A` is weighted lower than `C` here, which isn't what we want. I don't think we'd be able to work around this cleanly without switching to breadth-first traversal here.

Reply:
Okay, so... We want the length for the longest path to any node from root, ignoring any paths containing cycles.

Ok boi.

Reply:
> We want the length for the longest path to any node from root, ignoring any paths containing cycles.

This turns out to be easier to do with a DFS than a BFS. I wanted to avoid doing a BFS to avoid the O(n) space overhead of BFS. But... the current implementation needs to keep track of parents, which in turn becomes an O(n) space thing to do. :)

Fun.